### PR TITLE
Add function_id check in request controller

### PR DESCRIPTION
**Pull request by LuxoftAKutsan**
Request controller should not terminate request in case if function id
does not match.

Related issue :  APPLINK-28636

### CODE REVIEW

**Comment by LuxoftAKutsan** [comment]
@Abyzhynar, @agritsevich, @okozlovlux, @dev-gh, @dtrunov, @orhan-mehmedov Please review

File: src/components/application_manager/include/application_manager/application_manager_impl.h [hunk 1]
```diff
@@ -712,7 +712,9 @@ class ApplicationManagerImpl
    * @param connection_key - application id of request
    * @param corr_id correlation id of request
    */
-  void TerminateRequest(uint32_t connection_key, uint32_t corr_id) OVERRIDE;
+  void TerminateRequest(const uint32_t connection_key,
```
Comment (AByzhynar) [inline src/components/application_manager/include/application_manager/application_manager_impl.h]: @LuxoftAKutsan Update description accordingly

File: src/components/application_manager/include/application_manager/request_controller.h [hunk 2]
```diff
@@ -142,8 +142,9 @@ class RequestController {
   * @param force_terminate if true, request controller will terminate
   * even if not allowed by request
   */
-  void terminateRequest(const uint32_t& correlation_id,
-                        const uint32_t& connection_key,
+  void terminateRequest(const uint32_t correlation_id,
```
Comment (AByzhynar) [inline src/components/application_manager/include/application_manager/request_controller.h]: @LuxoftAKutsan Update description accordingly

Comment (AByzhynar) [inline src/components/application_manager/include/application_manager/request_controller.h]: @LuxoftAKutsan Please capitilize first letter on function name

File: src/components/application_manager/src/request_controller.cc [hunk 3]
```diff
@@ -227,37 +227,45 @@ void RequestController::removeNotification(
   LOG4CXX_DEBUG(logger_, "Cant find notification");
 }
 
-void RequestController::terminateRequest(const uint32_t& correlation_id,
-                                         const uint32_t& connection_key,
+void RequestController::terminateRequest(const uint32_t correlation_id,
```
Comment (AByzhynar) [inline src/components/application_manager/src/request_controller.cc]: @LuxoftAKutsan Please capitalize first letter on function name

File: src/components/application_manager/src/request_controller.cc [hunk 4]
```diff
@@ -227,37 +227,45 @@ void RequestController::removeNotification(
...
-    } else {
-      LOG4CXX_WARN(logger_, "Request was not terminated");
-    }
-    UpdateTimer();
+  if (!request) {
+    LOG4CXX_WARN(logger_, "Request not found in waiting_for_response");
+    return;
+  }
+  if (request->request()->function_id() != function_id) {
+    LOG4CXX_ERROR(logger_, "Request and response funciton_id's does not match");
```
Comment (AByzhynar) [inline src/components/application_manager/src/request_controller.cc]: @LuxoftAKutsan 
1)  `funciton_id's`->`function_id's`
2) `does not` -> `don't `

**Comment by AByzhynar** [comment]
@LuxoftAKutsan Did you check your build on your local PC? It does not compile.

File: src/components/application_manager/src/request_controller.cc (hunk 4 above)
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/request_controller.cc]: @AByzhynar what Apostrophe?

File: src/components/application_manager/include/application_manager/request_controller.h (hunk 2 above)
Comment (LuxoftAKutsan) [inline src/components/application_manager/include/application_manager/request_controller.h]: @AByzhynar see d40b5d9aadc17a3ff08bd10d1ee8b315234b4ba3

File: src/components/application_manager/src/request_controller.cc (hunk 3 above)
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/request_controller.cc]: @AByzhynar see d40b5d9aadc17a3ff08bd10d1ee8b315234b4ba3

File: src/components/application_manager/src/request_controller.cc (hunk 4 above)
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/request_controller.cc]: @AByzhynar see afbfe5a6b869d31e44b2ed57547d7cf3f875339f

File: src/components/application_manager/include/application_manager/request_controller.h [hunk 5]
```diff
@@ -152,16 +154,17 @@ class RequestController {
...
+                        const int32_t function_id);
 
   /**
   * @brief Removes request from queue
   *
   * @param mobile_correlation_id Active mobile request correlation ID
   *
   */
-  void OnHMIResponse(const uint32_t& correlation_id);
+  void OnHMIResponse(const uint32_t correlation_id, const int32_t function_id);
```
Comment (AByzhynar) [inline src/components/application_manager/include/application_manager/request_controller.h]: @LuxoftAKutsan Can `function_id` have negative value?

File: src/components/application_manager/src/application_manager_impl.cc [hunk 6]
```diff
@@ -1694,7 +1696,9 @@ bool ApplicationManagerImpl::ManageHMICommand(
     if (kResponse == message_type) {
       const uint32_t correlation_id =
           (*(message.get()))[strings::params][strings::correlation_id].asInt();
```
Comment (AByzhynar) [inline src/components/application_manager/src/application_manager_impl.cc]: @LuxoftAKutsan Why do we assign `int` value to `unsigned` type variable?

File: src/components/application_manager/src/application_manager_impl.cc [hunk 7]
```diff
@@ -1388,7 +1388,8 @@ void ApplicationManagerImpl::SendMessageToMobile(
   if (msg_to_mobile[strings::params].keyExists(strings::correlation_id)) {
     request_ctrl_.OnMobileResponse(
```
Comment (AByzhynar) [inline src/components/application_manager/src/application_manager_impl.cc]: @LuxoftAKutsan  ```
void OnMobileResponse(const uint32_t mobile_correlation_id,
                           const uint32_t& connection_key);
                           const uint32_t connection_key,
                           const int32_t function_id);
```
So why did you use `asInt()` for all params?

File: src/components/application_manager/include/application_manager/request_controller.h (hunk 5 above)
Comment (LuxoftAKutsan) [inline src/components/application_manager/include/application_manager/request_controller.h]: @AByzhynar function ID can be -1 (INVALID_VALUE) Please try to post all suggestions in one review iteration.

File: src/components/application_manager/src/application_manager_impl.cc (hunk 7 above)
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/application_manager_impl.cc]: @AByzhynar  see 9d172a2. Please try to post all review suggestions in one iteration.

**Pull request merged**